File: DLL/DLL.cpp
#include <iostream>
using namespace std;

// Node class
class Node {
public:
    int value;
    Node* next;
    Node* prev;

    Node(int val) {
        value = val;
        next = nullptr;
        prev = nullptr;
    }
};

// Doubly Linked List class
class DoublyLinkedList {
private:
    Node* head;
    Node* tail;
    int length;

public:
    DoublyLinkedList() {
        head = nullptr;
        tail = nullptr;
        length = 0;
    }

    // Insert at tail
    void push(int value) {
        Node* newNode = new Node(value);
        if (!head) {
            head = tail = newNode;
        } else {
            tail->next = newNode;
            newNode->prev = tail;
            tail = newNode;
        }
        length++;
    }

    // Remove from tail
    Node* pop() {
        if (!head) return nullptr;
        Node* poppedNode = tail;
        if (length == 1) {
            head = tail = nullptr;
        } else {
            tail = tail->prev;
            tail->next = nullptr;
            poppedNode->prev = nullptr;
        }
        length--;
        return poppedNode;
    }

    // Remove from head
    Node* shift() {
        if (!head) return nullptr;
        Node* oldHead = head;
        if (length == 1) {
            head = tail = nullptr;
        } else {
            head = oldHead->next;
            head->prev = nullptr;
            oldHead->next = nullptr;
        }
        length--;
        return oldHead;
    }

    // Insert at head
    void unshift(int value) {
        Node* newNode = new Node(value);
        if (length == 0) {
            head = tail = newNode;
        } else {
            head->prev = newNode;
            newNode->next = head;
            head = newNode;
        }
        length++;
    }

    // Access by index
    Node* get(int index) {
        if (index < 0 || index >= length) return nullptr;
        Node* current;
        if (index <= length / 2) {
            current = head;
            int count = 0;
            while (count != index) {
                current = current->next;
                count++;
            }
        } else {
            current = tail;
            int count = length - 1;
            while (count != index) {
                current = current->prev;
                count--;
            }
        }
        return current;
    }

    // Set value by index
    bool set(int index, int value) {
        Node* foundNode = get(index);
        if (foundNode != nullptr) {
            foundNode->value = value;
            return true;
        }
        return false;
    }

    // Insert at specific index
    bool insert(int index, int value) {
        if (index < 0 || index > length) return false;
        if (index == 0) {
            unshift(value);
            return true;
        }
        if (index == length) {
            push(value);
            return true;
        }

        Node* newNode = new Node(value);
        Node* beforeNode = get(index - 1);
        Node* afterNode = beforeNode->next;
        beforeNode->next = newNode;
        newNode->prev = beforeNode;
        newNode->next = afterNode;
        afterNode->prev = newNode;
        length++;
        return true;
    }

    // Remove at specific index
    Node* remove(int index) {
        if (index < 0 || index >= length) return nullptr;
        if (index == 0) return shift();
        if (index == length - 1) return pop();

        Node* removedNode = get(index);
        Node* beforeNode = removedNode->prev;
        Node* afterNode = removedNode->next;
        beforeNode->next = afterNode;
        afterNode->prev = beforeNode;
        removedNode->next = removedNode->prev = nullptr;
        length--;
        return removedNode;
    }

    // For debugging - print list forward
    void print() {
        Node* temp = head;
        while (temp != nullptr) {
            cout << temp->value << " <-> ";
            temp = temp->next;
        }
        cout << "NULL" << endl;
    }

    // Get length of list
    int size() {
        return length;
    }
};

File: DLL/DLL.js
// node class constructor
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

//doubly linked list class constructor
class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(value) { //inserting a node to the tail of the list
        let newNode = new Node(value);
        if (!this.head || this.length === 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length++;
        return this;
    }

    pop() { //removing a node from the tail of the list
        if (!this.head) return undefined;
        let poppedNode = this.tail;
        if (this.length === 1) {
            this.head = null;
            this.tail = null;
        } else {
            this.tail = poppedNode.prev;
            this.tail.next = null; //breaking the bond
            poppedNode.prev = null; //breaking the backward link
        }
        this.length--;
        return poppedNode;
    }

    shift() { //removing a node from the head of the list
        if (!this.head) return undefined;
        let oldHead = this.head;
        if (this.length === 1) {
            this.head = null;
            this.tail = null;
        } else {                              
            this.head = oldHead.next;
            this.head.prev = null;
            oldHead.next = null;
        }
        this.length--;
        return oldHead;                       
    }

    unshift(value) { //inserting a node to the head of the list
        let newNode = new Node(value);
        if (this.length === 0) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.head.prev = newNode;
            newNode.next = this.head;
            this.head = newNode;
        }
        this.length++;
        return this;
    }

    get(index) { //accessing a node by its position in DLL
        if (index < 0 || index >= this.length) return undefined;
        if (index <= this.length / 2) {
            let count = 0;
            let current = this.head;
            while(count != index) {
                current = current.next;
                count++;
            }
        } else {
            let count = this.length - 1;
            let current = this.tail;
            while (count != index) {
                current = current.prev;
                count--;
            }
        }
        return current;
    }

    set(index, value) { //replacing the value of a node in DLL
        let foundNode = this.get(index);
        if (foundNode != null) {
            foundNode.value = value;
            return true;
        }
        return false;
    }

    insert(index, value) { //adding a node in DLL by a certain position
        if (index < 0 || index > this.length) return false;
        if (index === 0) return this.unshift(value);
        if (index === this.length) return this.push(value);
        let insertedNode = new Node(value);
        let beforeNode = this.get(index - 1);
        let afterNode = beforeNode.next;
        beforeNode.next = insertedNode;
        insertedNode.prev = beforeNode;
        insertedNode.next = afterNode;
        afterNode.prev = insertedNode;
        this.length++;
        return true;
    }

    remove(index) { //removing a node in DLL by a certain position
        if (index < 0 || index >= this.length) return undefined;
        if (index === 0) return this.shift();
        if (index === this.length - 1) return this.pop();

        let removedNode = this.get(index);
        let beforeNode = removedNode.prev;
        let afterNode = removedNode.next;
        beforeNode.next = afterNode;
        afterNode.prev = beforeNode;
        afterNode = null;
        beforeNode = null;
        this.length--;
        return removedNode;
    }
}